File: src/webhook/currency.ts
import fetch from 'node-fetch';
import * as dayjs from 'dayjs';
import 'dayjs/locale/ru';

const currenciesUrl = 'https://www.cbr-xml-daily.ru/daily_json.js';
dayjs.locale('ru');

const randomCurrencies = [
  'AUD', 'AZN', 'AMD', 'BYN', 'BGN', 'BRL', 'HUF', 'HKD', 'DKK', 'INR', 'KZT', 'CAD', 'KGS', 'CNY', 'MDL', 'NOK',
  'PLN', 'RON', 'XDR', 'SGD', 'TJS', 'TRY', 'TMT', 'UZS', 'UAH', 'CZK', 'SEK', 'CHF', 'ZAR', 'KRW',
];

const getCurrencyData = (data, currencyKey) => {
  if (!data.Valute[currencyKey]) {
    return `[${currencyKey}] Нет данных по валюте!`;
  }
  const result = {
    cur: data.Valute[currencyKey].Value,
    isRise: data.Valute[currencyKey].Value > data.Valute[currencyKey].Previous ? 'опухло' : 'сдулось',
    change: Number((data.Valute[currencyKey].Value - data.Valute[currencyKey].Previous).toFixed(2)),
  };
  if (result.change < 0) {
    result.change = result.change * -1;
  }
  let changed = 'не изменилось';
  if (result.change !== 0) {
    changed = `${result.isRise} на ${result.change}`;
  }
  const name = data.Valute[currencyKey].Nominal > 1 ?
    `${data.Valute[currencyKey].Nominal} ${data.Valute[currencyKey].Name}` :
    data.Valute[currencyKey].Name;
  return `[${currencyKey}][${name}] ${result.cur} (${changed})`;
};

export default async function currency(currencyKey = null) {
  let data: any = await fetch(currenciesUrl);
  data = await data.json();
  if (
    !data.Valute ||
    !data.Valute.USD ||
    !data.Valute.EUR ||
    !data.Valute.USD.Value ||
    !data.Valute.USD.Previous ||
    !data.Valute.EUR.Value ||
    !data.Valute.EUR.Previous
  ) {
    return 'Не могу получить данные с cbr-xml-daily ;(';
  }
  const dates = {
    updated: dayjs(data.Date || 0).format('DD.MM.YYYY (HH:mm)'),
    prevDate: dayjs(data.PreviousDate || 0).format('DD.MM.YYYY (HH:mm)'),
  };
  let currencies;
  if (currencyKey) {
    currencies = `${getCurrencyData(data, currencyKey)}`;
  } else {
    currencies = `${getCurrencyData(data, 'USD')}
${getCurrencyData(data, 'EUR')}
${getCurrencyData(data, 'GBP')}
${getCurrencyData(data, 'JPY')}
${getCurrencyData(data, randomCurrencies[Math.floor(Math.random() * randomCurrencies.length)])}`;
  }
  return `[Обновлено: ${dates.updated} | Предыдущие данные: ${dates.prevDate}]

${currencies}

(Источник: cbr-xml-daily.ru)
`;
}

File: internal/commands/currency/currency.go
package currency

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"math/rand"
	"net/http"
	"strings"
	"time"
)

const currenciesURL = "https://www.cbr-xml-daily.ru/daily_json.js"

var randomCurrencies = []string{
	"AUD", "AZN", "AMD", "BYN", "BGN", "BRL", "HUF", "HKD", "DKK", "INR", "KZT", "CAD", "KGS", "CNY", "MDL", "NOK",
	"RON", "XDR", "SGD", "TJS", "TRY", "TMT", "UZS", "UAH", "CZK", "SEK", "ZAR", "KRW",
}

type cbrResponse struct {
	Date         string `json:"Date"`
	PreviousDate string `json:"PreviousDate"`
	Valute       map[string]struct {
		Name     string  `json:"Name"`
		Nominal  float64 `json:"Nominal"`
		Value    float64 `json:"Value"`
		Previous float64 `json:"Previous"`
	} `json:"Valute"`
}

func Format(ctx context.Context, currencyKey string) (string, error) {
	data, err := fetch(ctx)
	if err != nil {
		return "", err
	}
	if data.Valute == nil ||
		data.Valute["USD"].Value == 0 ||
		data.Valute["EUR"].Value == 0 {
		return "Не могу получить данные с cbr-xml-daily ;(", nil
	}

	updated := formatCBRDate(data.Date)
	prevDate := formatCBRDate(data.PreviousDate)

	var currencies string
	if currencyKey != "" {
		currencies = getCurrencyData(data, currencyKey)
	} else {
		randomKey := randomCurrencies[rand.Intn(len(randomCurrencies))]
		currencies = strings.Join([]string{
			getCurrencyData(data, "USD"),
			getCurrencyData(data, "EUR"),
			getCurrencyData(data, "RSD"),
			getCurrencyData(data, "GBP"),
			getCurrencyData(data, "GEL"),
			getCurrencyData(data, "CHF"),
			getCurrencyData(data, "PLN"),
			getCurrencyData(data, randomKey),
		}, "\n")
	}

	return fmt.Sprintf("[Обновлено: %s | Предыдущие данные: %s]\n\n%s\n\n(Источник: cbr-xml-daily.ru)\n",
		updated, prevDate, currencies), nil
}

func fetch(ctx context.Context) (*cbrResponse, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, currenciesURL, nil)
	if err != nil {
		return nil, err
	}
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	var data cbrResponse
	if err := json.Unmarshal(body, &data); err != nil {
		return nil, err
	}
	return &data, nil
}

func getCurrencyData(data *cbrResponse, currencyKey string) string {
	val, ok := data.Valute[currencyKey]
	if !ok {
		return fmt.Sprintf("[%s] Нет данных по валюте!", currencyKey)
	}
	change := val.Value - val.Previous
	if change < 0 {
		change = -change
	}
	isRise := "сдулось"
	if val.Value > val.Previous {
		isRise = "опухло"
	}
	changed := "не изменилось"
	if change != 0 {
		changed = fmt.Sprintf("%s на %.2f", isRise, change)
	}
	name := val.Name
	if val.Nominal > 1 {
		name = fmt.Sprintf("%.0f %s", val.Nominal, val.Name)
	}
	return fmt.Sprintf("[%s][%s] %v (%s)", currencyKey, name, val.Value, changed)
}

func formatCBRDate(raw string) string {
	t, err := time.Parse(time.RFC3339, raw)
	if err != nil {
		return raw
	}
	loc, _ := time.LoadLocation("Europe/Moscow")
	return t.In(loc).Format("02.01.2006 (15:04)")
}
